### ISSUE TITLE
Automatic streaming

### ISSUE THREAD

**Issue by duncanwilcox** [Selz/plyr#289]
The grid of browser support, desired capabilities (encryption, subtitles) and support library tradeoffs is pretty complex. I think it would great if plyr.io could pick the best option by detecting what the client capabilities are and what the desired features are, picking the appropriate streaming technology and client support library.

For example:

- for older browsers that don't have MSE a straight mp4 is the only option
- iOS only supports HLS, and no support library is required
- HLS is broken on Chrome on Android 4+
- if MSE is supported, DASH is probably preferable anyway because HLS uses the MPEG-2 Transport Stream that's higher overhead compared to the straight mp4 or webm, supported by DASH
- Shaka is about a third of the size of the dashif dash.js, but I'm not familiar with what differences there might be

There are other capabilities such as ad insertion, DRM and more things which I'm not familiar with, and I believe DASH is a preferable format there as well.

This doesn't address what needs to be done on the server side, but it's not a plyr.io concern anyway.

Are #39 and #159 still relevant?

**Comment by SamPotts** [comment]
It's a tough one. I'd like to add something like this into the library but I have concerns over the size of Plyr getting larger and more complex. I recently added some notes and examples and about [streaming support](https://github.com/Selz/plyr#streaming) but it's something that is new to me at this point. I'm aware that support is all over the place at the moment which makes it even harder.

**Comment by duncanwilcox** [comment]
Right, this was precisely building on that. I think you're 90% there, I don't think there's any need to support HLS directly because if the browser supports it then it's already there, if it doesn't then it's definitely preferable to look at the more modern DASH.

DASH on the other hand is a pretty complex standard, I definitely think you can lean on Google's work on Shaka (barring any licensing issues).

https://github.com/Dash-Industry-Forum/dash.js/wiki/Meeting-Minutes suggests

> Shaka also supports a subset of DASH (no emsg support, no multi-period).

which I'm unfamiliar with but something that could determine a switch to the dashif library.

The plyr.io API would let the client specify a list of sources and feature requirements, and determine what is required from that.

For example the player initializer might specify an mp4 URL, a m3u8 HLS URL, and an mpd URL, and on iOS this would result in the m3u8 being set as the video src, while on an MSE-enabled browser would result in loading Shaka and initializing it with the mpd.

Frankly a lot of the additional features supported by DASH seem very specific (https://bitmovin.com/mpeg-dash-vs-apple-hls-vs-microsoft-smooth-streaming-vs-adobe-hds/ has a comparison table), so the above might be a good start.

**Comment by zamirkhan** [comment]
I'm currently using plyr and interested in adding streaming support (HLS, DASH & Shaka) for as many clients as possible. I see the individual examples linked from the readme and they are helpful. The last part that I need, which I presume this issue refers to, is the ability to select and provide the right source to the right client. If I understand correctly, this is not currently built into plyr. However, is there an example somewhere of how one could implement this decision themselves and thus supply the "correct" source to the plyr player?

**Comment by jhovad** [comment]
@zamirkhan I got it from there: https://transloadit.com/demos/video-encoding/compatibility-in-2018/ ... I think it is not so difficult logic (if hls supported and hls source exist ..., elseif dash supported and dash source exist ..., else plain ...). Maybe some small script in form of Plyr plugin would be very helpful for current times of this video mess.
